### PR TITLE
[Daml] Handle new Numeric builtins more nicely

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
In this PR, we rewrite the change from #16852 without changing Daml stdlib signature (we just add more constraint on type) and 
handle properly `Additive` and `Multiplicative` type class form Numeric that was not handle at all.

### CODE REVIEW

File: compiler/damlc/daml-prim-src/GHC/Classes.daml [hunk 1]
```diff
@@ -407,46 +419,95 @@ class NumericScale (n : Nat) where
...
+--    -- | HIDE
+--    textToNumericBuiltin: Text -> Optional (Numeric n)
+    -- | HIDE
+#ifdef DAML_BIGNUMERIC
+    bigNumericToNumericBuiltin: BigNumeric -> Numeric n
+#endif
 
+    addNumericBuiltin = primitive @"BEAddNumeric"
+    subNumericBuiltin = primitive @"BEAddNumeric"
+    roundNumericBuiltin = primitive @"BERoundNumeric"
```
Comment (remyhaemmerle-da) [inline compiler/damlc/daml-prim-src/GHC/Classes.daml:449]: We put those 3 builtin together the one that require a witness for the sake of consistency

File: compiler/damlc/daml-prim-src/GHC/Classes.daml [hunk 2]
```diff
@@ -445,6 +451,58 @@ instance NumericScale 34 where numericScale _ = 34; numericScalePrivate _ = ()
 instance NumericScale 35 where numericScale _ = 35; numericScalePrivate _ = ()
 instance NumericScale 36 where numericScale _ = 36; numericScalePrivate _ = ()
 instance NumericScale 37 where numericScale _ = 37; numericScalePrivate _ = ()
+#else
+class NumericScale (n : Nat) where
+    -- | Get the scale of a `Numeric` as an integer. For example,
+    -- `numericScale (3.14159 : Numeric 5)` equals `5`.
+    numericScale : Int
```
Comment (remyhaemmerle-da) [inline compiler/damlc/daml-prim-src/GHC/Classes.daml:458]: we do not need proxy as only ground instances are defined.

File: compiler/damlc/daml-prim-src/GHC/Num.daml [hunk 3]
```diff
@@ -102,6 +102,27 @@ class Signed a where
     -- | The absolute value: that is, the value without the sign.
     abs : a -> a
 
+-- | Use the `Divisible` class for types that can be divided.
```
Comment (remyhaemmerle-da) [inline compiler/damlc/daml-prim-src/GHC/Num.daml:105]: boy scout cleaning. We define all the classes before the instances.

File: compiler/damlc/daml-stdlib-src/DA/Internal/Prelude.daml [hunk 4]
```diff
@@ -562,97 +562,66 @@ fst (x, _) = x
...
+  -- We first use bankers rounding and calculate the difference
+  -- we idendity the problematic rounding by multiply 2 and checking idenmpotency by bankers rounding and its sign
+   let
+      t = roundBankers d x
+      diff = t - x
+      diffx2 = diff * intToNumeric 2
+    in
+      if diffx2 == roundBankers d diffx2 && diff * signum x < intToNumeric 0
+      then x - diff
+      else t
```
Comment (remyhaemmerle-da) [inline compiler/damlc/daml-stdlib-src/DA/Internal/Prelude.daml:600]: We simplify the implementation to not need `divNumeric` nor exponentiation.

**Review by samuel-williams-da — APPROVED** [review]
Looks good!
As for feature flag, you're already doing the right thing `DAML_NAT_TYPE_ERASURE` comes from the `featureNatTypeErasure` that Dylan added 2 months ago, I'm not sure what else you would want to do here.
And finally data deps - do you know what breaks here? Theres reference to Numeric and Nat in there, but just for src code conversion, I wouldn't expect it to break

File: compiler/damlc/daml-stdlib-src/DA/Internal/Prelude.daml (hunk 4 above)
Comment (samuel-williams-da) [reply compiler/damlc/daml-stdlib-src/DA/Internal/Prelude.daml:600]: A nice simplification :)

**Review by cocreature — COMMENTED** [review]
looks reasonable at a high level,

data-dependencies are probably mostly fine since there are no user defined instances of these classes. The one thing I could see breaking is the ability to call a method defined in an older version with a NumericScale n constraint.

File: compiler/damlc/daml-stdlib-src/DA/Internal/Prelude.daml [hunk 5]
```diff
@@ -562,97 +562,66 @@ fst (x, _) = x
...
-      -- This is defined in DA.Numeric but we inline it here to avoid creating a cyclic module dependency.
-      divNumeric : Numeric n1 -> Numeric n2 -> Numeric n
-      divNumeric = primitive @"BEDivNumericLegacy"
-      problematicDiff = (0.5 : Numeric 1) `divNumeric` (10.0 ^ d : Numeric 0) * signum x
-  in if diff == problematicDiff
-       then x + problematicDiff
-       else t
-#endif
+  -- We first use bankers rounding and calculate the difference
+  -- we idendity the problematic rounding by multiply 2 and checking idenmpotency by bankers rounding and its sign
```
[the remark this answers is not in the document]
Comment (dylant-da) [reply compiler/damlc/daml-stdlib-src/DA/Internal/Prelude.daml:592]:
```suggestion
  -- we identify the problematic rounding by multiply 2 and checking idempotency by bankers rounding and its sign
```

**Review by dylant-da — APPROVED** [review]
Looks like what we discussed, thanks!

**Pull request merged**